[inductor] New approach for computing triton load/store masks

fdrocha opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89593
* __->__ #89566

This PR changes the way masks for loads/stores are computed in triton backend of inductor.

New approach is to iterate over all variables used in indexing expression and add the corresponding mask variables to the set that will be used. For indexing variables like `x0`, `y1` and  `r3` it adds `xmask`, `ymask` and `rmask` respectively.
For indexing variables like `tmp5` (i.e., indirect indexing), it uses the new `mask_vars` attribute of the corresponding `TritonCSEVariable` object, which is populated when variable is created.

I started working on this with the aim of fixing https://github.com/pytorch/torchdynamo/issues/1654, which meanwhile was fixed by #89524 with a different approach, making this change less necessary. However note that #89524 fixes the issue by broadcasting the indices that are being loaded to a larger size, while this approach fixes it by making the mask have only the necessary terms.

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire @lezcano @ngimel

Commit:
[inductor] New approach for indirect indexing mask

[ghstack-poisoned]

jansel commented on torch/_inductor/codegen/triton.py:
```diff
@@ -722,65 +723,56 @@ def indexing(
         have_loop_vars = False
-        mask = []
-        dense_mask = []
+        dense_mask = set()
```
Python sets have nondeterministic order.  If we use them we will need to call sort later.

jansel commented on torch/_inductor/codegen/triton.py:
```diff
@@ -722,65 +723,56 @@ def indexing(
-            mask = ["None"]
-
-        return index_str, " & ".join(map(str, mask))
+        mask_str = " & ".join(map(str, mask_vars)) if mask_vars else "None"
```
```suggestion
        mask_str = " & ".join(map(str, sorted(mask_vars))) if mask_vars else "None"
```
Not sure if there are other places this will be needed too.

fdrocha replied:
Thanks, I believe this is the only one.

jansel reviewed (changes requested):
Any performance results from this change?

Add a test?

Commit:
Update on "[inductor] New approach for computing triton load/store masks"


This PR changes the way masks for loads/stores are computed in triton backend of inductor.

New approach is to iterate over all variables used in indexing expression and add the corresponding mask variables to the set that will be used. For indexing variables like `x0`, `y1` and  `r3` it adds `xmask`, `ymask` and `rmask` respectively.
For indexing variables like `tmp5` (i.e., indirect indexing), it uses the new `mask_vars` attribute of the corresponding `TritonCSEVariable` object, which is populated when variable is created.

I started working on this with the aim of fixing https://github.com/pytorch/torchdynamo/issues/1654, which meanwhile was fixed by #89524 with a different approach, making this change less necessary. However note that #89524 fixes the issue by broadcasting the indices that are being loaded to a larger size, while this approach fixes it by making the mask have only the necessary terms.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire lezcano ngimel 

[ghstack-poisoned]

fdrocha commented:
> Any performance results from this change?

Will get some benchmarks soon.

> Add a test?

https://github.com/pytorch/pytorch/pull/89524 added a test to check for regressions in https://github.com/pytorch/torchdynamo/issues/1654, that is now merged to master. Anything else that you think would make sense to test?

Commit:
Update on "[inductor] New approach for computing triton load/store masks"


This PR changes the way masks for loads/stores are computed in triton backend of inductor.

New approach is to iterate over all variables used in indexing expression and add the corresponding mask variables to the set that will be used. For indexing variables like `x0`, `y1` and  `r3` it adds `xmask`, `ymask` and `rmask` respectively.
For indexing variables like `tmp5` (i.e., indirect indexing), it uses the new `mask_vars` attribute of the corresponding `TritonCSEVariable` object, which is populated when variable is created.

I started working on this with the aim of fixing https://github.com/pytorch/torchdynamo/issues/1654, which meanwhile was fixed by #89524 with a different approach, making this change less necessary. However note that #89524 fixes the issue by broadcasting the indices that are being loaded to a larger size, while this approach fixes it by making the mask have only the necessary terms.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire lezcano ngimel 

[ghstack-poisoned]

Commit:
Update on "[inductor] New approach for computing triton load/store masks"


This PR changes the way masks for loads/stores are computed in triton backend of inductor.

New approach is to iterate over all variables used in indexing expression and add the corresponding mask variables to the set that will be used. For indexing variables like `x0`, `y1` and  `r3` it adds `xmask`, `ymask` and `rmask` respectively.
For indexing variables like `tmp5` (i.e., indirect indexing), it uses the new `mask_vars` attribute of the corresponding `TritonCSEVariable` object, which is populated when variable is created.

I started working on this with the aim of fixing https://github.com/pytorch/torchdynamo/issues/1654, which meanwhile was fixed by #89524 with a different approach, making this change less necessary. However note that #89524 fixes the issue by broadcasting the indices that are being loaded to a larger size, while this approach fixes it by making the mask have only the necessary terms.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire lezcano ngimel 

[ghstack-poisoned]

Commit:
Update on "[inductor] New approach for computing triton load/store masks"


This PR changes the way masks for loads/stores are computed in triton backend of inductor.

New approach is to iterate over all variables used in indexing expression and add the corresponding mask variables to the set that will be used. For indexing variables like `x0`, `y1` and  `r3` it adds `xmask`, `ymask` and `rmask` respectively.
For indexing variables like `tmp5` (i.e., indirect indexing), it uses the new `mask_vars` attribute of the corresponding `TritonCSEVariable` object, which is populated when variable is created.

I started working on this with the aim of fixing https://github.com/pytorch/torchdynamo/issues/1654, which meanwhile was fixed by #89524 with a different approach, making this change less necessary. However note that #89524 fixes the issue by broadcasting the indices that are being loaded to a larger size, while this approach fixes it by making the mask have only the necessary terms.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire lezcano ngimel 

[ghstack-poisoned]

fdrocha commented:
Ok finally got some performance results for the change (setting up torchbenchmark for the first time took longer than I expected). TLDR: minimal impact.

Here are the summary results for the CUDA torchbench inference tasks:

```
Before change:
speedup             gmean=1.42x mean=1.84x
abs_latency         gmean=nanx mean=78.14x
compilation_latency mean=5.6 seconds
compression_ratio   mean=1.2x

After change:
speedup             gmean=1.40x mean=1.80x
abs_latency         gmean=nanx mean=72.97x
compilation_latency mean=5.7 seconds
compression_ratio   mean=1.1x
```

I think it is still worth making the change, but let me know what you think @jansel

Commit:
Update on "[inductor] New approach for computing triton load/store masks"


This PR changes the way masks for loads/stores are computed in triton backend of inductor.

New approach is to iterate over all variables used in indexing expression and add the corresponding mask variables to the set that will be used. For indexing variables like `x0`, `y1` and  `r3` it adds `xmask`, `ymask` and `rmask` respectively.
For indexing variables like `tmp5` (i.e., indirect indexing), it uses the new `mask_vars` attribute of the corresponding `TritonCSEVariable` object, which is populated when variable is created.

I started working on this with the aim of fixing https://github.com/pytorch/torchdynamo/issues/1654, which meanwhile was fixed by #89524 with a different approach, making this change less necessary. However note that #89524 fixes the issue by broadcasting the indices that are being loaded to a larger size, while this approach fixes it by making the mask have only the necessary terms.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire lezcano ngimel 

[ghstack-poisoned]

jansel commented:
> Ok finally got some performance results for the change (setting up torchbenchmark for the first time took longer than I expected). TLDR: minimal impact.
> 
> Here are the summary results for the CUDA torchbench inference tasks:
> 
> ```
> Before change:
> speedup             gmean=1.42x mean=1.84x
> abs_latency         gmean=nanx mean=78.14x
> compilation_latency mean=5.6 seconds
> compression_ratio   mean=1.2x
> 
> After change:
> speedup             gmean=1.40x mean=1.80x
> abs_latency         gmean=nanx mean=72.97x
> compilation_latency mean=5.7 seconds
> compression_ratio   mean=1.1x
> ```
> 
> I think it is still worth making the change, but let me know what you think @jansel

These numbers look like a slowdown?  Have you figured out why?

fdrocha commented:
> These numbers look like a slowdown? Have you figured out why?

You're right, I had misread the abs_latency mean. It seems the benchmark results I'm getting are not very stable over different runs, trying to figure out why...

ngimel commented:
Yeah that's strange, I wouldn't expect slow-downs from these changes. If result differences persist, try finding some model with a particularly big impact and take a look at it.

fdrocha commented:
> Yeah that's strange, I wouldn't expect slow-downs from these changes. If result differences persist, try finding some model with a particularly big impact and take a look at it.

Yes, it's taking me way longer than I expected to set up a machine to run benchmarks properly. Think I'm almost there and will repost benchmarks.

fdrocha commented:
Ok, finally got the benchmarks work consistently and, as you'd expect, you don't really see any differences. 

These are for the pytorch models, in inference mode (command was `python benchmarks/dynamo/torchbench.py -dcuda --float32 -n25 --performance --fast --inductor --filter '^pytorch_'`):

Before change:

```
speedup             gmean=1.38x mean=1.40x
abs_latency         gmean=9.23x mean=18.80x
compilation_latency mean=2.4 seconds
compression_ratio   mean=1.2x
```

After change:

```
speedup             gmean=1.38x mean=1.40x
abs_latency         gmean=9.27x mean=18.84x
compilation_latency mean=2.3 seconds
compression_ratio   mean=1.2x
```

@jansel @ngimel let me know what you think.

ngimel reviewed (approved):
The benchmarking results look good

fdrocha commented:
@pytorchbot merge -gr

Commit:
Update on "[inductor] New approach for computing triton load/store masks"


This PR changes the way masks for loads/stores are computed in triton backend of inductor.

New approach is to iterate over all variables used in indexing expression and add the corresponding mask variables to the set that will be used. For indexing variables like `x0`, `y1` and  `r3` it adds `xmask`, `ymask` and `rmask` respectively.
For indexing variables like `tmp5` (i.e., indirect indexing), it uses the new `mask_vars` attribute of the corresponding `TritonCSEVariable` object, which is populated when variable is created.

I started working on this with the aim of fixing https://github.com/pytorch/torchdynamo/issues/1654, which meanwhile was fixed by #89524 with a different approach, making this change less necessary. However note that #89524 fixes the issue by broadcasting the indices that are being loaded to a larger size, while this approach fixes it by making the mask have only the necessary terms.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire lezcano ngimel 

[ghstack-poisoned]

fdrocha commented:
@pytorchbot merge -f 'Test failure is CI issue'

clee2000 commented:
@pytorchbot revert -m "broke test_invalid_operand_issue1_cuda in inductor/test_torchinductor on https://github.com/pytorch/pytorch/actions/runs/3657444733/jobs/6181700572" -c ignoredsignal
